pennylane/labs/dla/dense_util.py

```diff
@@ -0,0 +1,413 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utility tools for dense Lie algebra representations"""
+# pylint: disable=possibly-used-before-assignment
+from itertools import combinations_with_replacement
+from typing import Iterable, Optional, Union
+
+import numpy as np
+from scipy.linalg import sqrtm
+
+import pennylane as qml
+from pennylane.operation import Operator
+from pennylane.ops.qubit.matrix_ops import _walsh_hadamard_transform
+from pennylane.pauli import PauliSentence, PauliVSpace, PauliWord
+from pennylane.typing import TensorLike
+
+
+def _make_phase_mat(n: int) -> np.ndarray:
+    r"""Create an array with shape ``(2**n, 2**n)`` containing powers of :math:`i`.
+    For the entry at position ``(i, k)``, the entry is ``1j**p(i,k)`` with the
+    power being the Hamming weight of the product of the binary bitstrings of
+    ``i`` and ``k``.
+    """
+    _slice = (slice(None), None, None)
+    # Compute the bitwise and of the row and column index for all matrix entries
+    ids_bit_and = np.bitwise_and(np.arange(2**n)[:, None], np.arange(2**n)[None])
+    # Compute the Hamming weight of the bitwise and. Can be replaced by bitwise_count with np>=2.0
+    hamming_weight = np.sum(
+        np.bitwise_and(ids_bit_and[None] >> np.arange(n + 1)[_slice], 1), axis=0
+    )
+    return 1j**hamming_weight
+
+
+def _make_permutation_indices(dim: int) -> list[np.ndarray]:
+    r"""Make a list of ``dim`` arrays of length ``dim`` containing the indices
+    ``0`` through ``dim-1`` in a specific permutation order to match the Walsh-Hadamard
+    transform to the Pauli decomposition task."""
+    indices = [qml.math.arange(dim)]
+    for idx in range(dim - 1):
+        indices.append(qml.math.bitwise_xor(indices[-1], (idx + 1) ^ (idx)))
+    return indices
+
+
+def _make_extraction_indices(n: int) -> tuple[tuple]:
+    r"""Create a tuple of two tuples of indices to extract Pauli basis coefficients.
+    The first tuple of indices as bit strings encodes the presence of a Pauli Z or Pauli Y
+    on the ``k``\ th wire in the ``k``\ th bit. The second tuple encodes the presence of
+    Pauli X or Pauli Y. That is, for a given position, the four different Pauli operators
+    are encoded as ``(0, 0) = "I"``, ``(0, 1) = "X"``, ``(1, 0) = "Z"`` and ``(1, 1) = "Y"``.
+    """
+    if n == 1:
+        return ((0, 0, 1, 1), (0, 1, 1, 0))
+
+    ids0, ids1 = np.array(_make_extraction_indices(n - 1))
+    return (
+        tuple(np.concatenate([ids0, ids0, ids0 + 2 ** (n - 1), ids0 + 2 ** (n - 1)])),
+        tuple(np.concatenate([ids1, ids1 + 2 ** (n - 1), ids1 + 2 ** (n - 1), ids1])),
+    )
+
+
+def pauli_coefficients(H: TensorLike) -> np.ndarray:
+    r"""Computes the coefficients of a Hermitian matrix in the Pauli basis.
+
+    The coefficients are ordered lexicographically in the Pauli group.
+    I.e. for ``n=2`` qubits we have the following ordering.
+
+    .. code-block::
```

Comment:
@dwierichs is this correct?

Reply:
Mhm. No that looks like a super randomly ordered list o.O
Am I being confused here? :sweat_smile:

Reply:
Ah, it is the canonical ordering if you use stabilizer tableaux, and it is the order returned by `qml.pauli.pauli_group`! But we don't use that here.

Reply:
Yep exactly, its what we get from `pauli_group`, I thought this is canonical / what we use here 😅

Reply:
Mhm, no I think we return things in the lexicographical ordering, which usually is `["III", "IIX", "IIY", "IIZ", "IXI", ...]`

Reply:
Updated it with that, thanks! Wasnt aware!